docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,208 @@
+        Value: foo
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 15s
+      idleTimeout: 20s
+    Origin: [ ... ]
+  - Subset:
+      - Key: kuma.io/service
+        Not: false
+        Value: bar
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 20s
+      idleTimeout: 20s
+    Origin: [ ... ]
+```
+
+This structure is used by both "from" and "to" policies plugins.
+Policy plugins use the `Rules` list to compute `Conf` for the known destination.
+In example, if we want to know the `Conf` for outbound with `kuma.io/service: bar` tags,
+we're going to iterate over `Rules` and check if our outbound belongs to the `Rules[].Subset`.
+
+The approach has a number of disadvantages:
+
+* even though it works well for "from" policies, using it for "to" policies doesn't feel natural
+  * there is no concept of "subsets" in outbounds, we have to artificially build `{kuma.io/service: $svc}` subset
+  * subset-based matching is build to handle subset overlapping scenarios, but it's never happening in outbounds
+* when referencing the real MeshHTTPRoute we had to use subset with artificial `__rule-matches-hash__` key
+* adding support for `MeshService` forces us to introduce another artificial key like `__resource-name__`
+* Inspect API with subsets is tricky when referencing real resources (like MeshHTTPRoute or MeshService)
+
+We need a new format that'd make things easier when it comes to referencing real resources in `to[]`.
+
+## Considered Options and Decision Outcome
+
+No changes for "from" policies, both policy matching algorithm and Inspect API stays the same.
+
+For "to" policies the algorithm is going to return the additional structure:
+
+```go
+type ToRules struct {
+    Rules         Rules 
+    ResourceRules map[UniqueResourceKey]ResourceRule // new field
+}
+
+// UniqueResourceKey is a way to uniquely identify a resource, should include ResourceType, Name and Mesh
+type UniqueResourceKey struct{}
+
+type ResourceRule struct {
+    Resource core_model.ResourceMeta
+    Conf     interface{}
+    Origin   []core_model.ResourceMeta
+}
+```
+
+For a simple targetRef with `name/namespace`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+      namespace: finance
+    conf: conf1
+```
+
+we're going to produce a single `ResourceRule` entity:
+
+```yaml
+resourceRules:
+  - resource:
+      name: backend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+```
+
+When targetRef has kind `MeshService` and uses `labels` we're going to produce multiple `ResourceRule` entities
+for each real `MeshService` that satisfies `labels`, i.e for the given `targetRef`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      labels:
+        k8s.kuma.io/namespace: finance
+    conf: conf1
+```
+
+we are going to fetch all `MeshService` resources, filter them by `k8s.kuma.io/namespace: finance` and produce N rules:
+
+```yaml
+resourceRules:
```

Comment:
> Won't it be unexpected that the response changes based on what services are running at the moment?

Not "running", but on "what services are defined". Response change depending on "what services are running" wouldn't be nice I guess – you lost connection between DPP and CP and the Inspect API response changed. But response change depending on "what services are defined" I think is more than acceptable, no?

Reply:
yeah definitely, by running i mean in the sense of there's a Service that exists.

I don't know. the comparison with the example I gave for a specific service still stands I think.

maybe it depends on what you're using the API for. I mean in a sense you kind of lose the information that the labels give you. Which is "all services that fulfill these labels" vs "this list of specific services"

Reply:
yeah I guess would be nicer to show both labels and the actual MeshServices that were selected by these labels. So the ResourceRule might looks something like:
```go
type ResourceRule struct {
    Resource         core_model.ResourceMeta
    Conf             interface{}
    Origin           []core_model.ResourceMeta
    SelectedByLabels map[string]string // optional field that's set when `targetRef.Labels` were specified
}
```
wdyt?

Reply:
Imagine user created a policy:
```yaml
kind: MeshTimeout
metadata:
  namespace: finance
spec:
  to:
    - targetRef:
        kind: MeshService
        labels:
          k8s.kuma.io/namespace: finance
        default: conf1
    - targetRef:
        kind: MeshService
        labels:
          kuma.io/zone: zone-1
        default: conf2
```
how the list of Rules should look like in the Inspect API in case we don't want to use specific MeshServices?

Reply:
`SelectedByLabels` is kind of similar to Origin. So I decided to extend the Origin type with the index of to-item in the `to[]` array that contributed to the conf. This also solves the problem when the final conf for `MeshService` is affected by to-item with `Mesh`:

```yaml
kind: MeshTimeout
metadata:
  namespace: finance
  name: mt1
spec:
  to:
    - targetRef:
        kind: Mesh
        default: conf1
    - targetRef:
        kind: MeshService
        name: backend
        default: conf2
```
the rule will look like:

```yaml
resourceRules:
  meshservice.mesh-1.backend:
    resource:
      name: backend
      namespace: finance
    conf: merge(conf2, conf1)
    origin:
      - resource:
          name: mt1
          namespace: finance
        toIdx: 0
      - resource:
          name: mt1
          namespace: finance
        toIdx: 1
```
I updated the madr according to this.